### PR TITLE
Refine path section in message serialization doc

### PR DESCRIPTION
**Pull request by zhenlineo**

### CODE REVIEW

File: community/bolt/src/docs/dev/serialization.asciidoc [hunk 1]
```diff
@@ -407,8 +408,9 @@ Relationship (signature=0x52) {
 
 [[bolt-value-pathstruct]]
 ==== Path
-A Path represents a walk through a graph.
-Its representation consists of a set of distinct nodes included within the path, a similar set of distinct relationships and a sequence describing the shape of the path.
+A Path is a sequence of alternating nodes and relationships corresponding to a walk in the graph.
+The sequence always begins and ends with a node.
```
Comment (petraselmer) [inline community/bolt/src/docs/dev/serialization.asciidoc]: The sequence -> The path

File: community/bolt/src/docs/dev/serialization.asciidoc [hunk 2]
```diff
@@ -407,8 +408,9 @@ Relationship (signature=0x52) {
...
+Its representation consists of a list of distinct nodes, a list of distinct relationships and a sequence of integer indexes describing the shape of the path.
```
Comment (petraselmer) [inline community/bolt/src/docs/dev/serialization.asciidoc]: I would suggest "integer indexes" -> "integers". 
Also, "describing the shape of the path" -> "describing the path traversal"

File: community/bolt/src/docs/dev/serialization.asciidoc [hunk 3]
```diff
@@ -420,6 +422,46 @@ Path (signature=0x50) {
 }
 ----
 
+The nodes (`N`) and relationships (`R`) are defined as follows:
```
Comment (petraselmer) [inline community/bolt/src/docs/dev/serialization.asciidoc]: Specify precisely what `N` and `R` are; e.g. something like "The two dictionaries, `N` and `R` are defined as follows" (the contents therein are specified later, so no need to say nodes and relationships in this sentence)

File: community/bolt/src/docs/dev/serialization.asciidoc [hunk 4]
```diff
@@ -420,6 +422,46 @@ Path (signature=0x50) {
...
+
+- `N` contains all the unique nodes in the path
+- `R` contains all the unique relationships in the path
+- For `N`, the index is an integer commencing with 0 and incrementing by 1
+- For `R`, the index is an integer commencing with 1 and incrementing by 1
+- The value component for both `N` and `R` is the data corresponding to the node or relationship (as defined above)
```
Comment (petraselmer) [inline community/bolt/src/docs/dev/serialization.asciidoc]: "(as defined above)"  --> where? I think it is better to list the contents here for ease of readability.

File: community/bolt/src/docs/dev/serialization.asciidoc [hunk 5]
```diff
@@ -420,6 +422,46 @@ Path (signature=0x50) {
...
+- In `N`, the first element must _always_ be the first node in the path (thus having 0 as the key)
```
Comment (petraselmer) [inline community/bolt/src/docs/dev/serialization.asciidoc]: Here, "key" has been used, whereas earlier "index" was used (lines 429 and 430), so inconsistent terminology. This ought to be fixed... maybe "index" is better, as it marries well with the paragraphs that follow...

File: community/bolt/src/docs/dev/serialization.asciidoc [hunk 6]
```diff
@@ -420,6 +422,46 @@ Path (signature=0x50) {
...
+- No other _explicit_ rules apply as to either (i) the ordering of the other nodes in `N`, or (ii) the ordering of any of the relationships in `R`. However, while not required, it is recommended that implementations aim to list entities (i.e. nodes and relationships) in the order in which they are first encountered while traversing the path. This may help with the efficiency of reading and writing
+
+In the context of a transmission, the path is represented as a sequence of integers; we define `S` as the transmitted sequence, and `S'` as the full sequence.
```
Comment (petraselmer) [inline community/bolt/src/docs/dev/serialization.asciidoc]: Has the concept "transmission" been defined previously? I would suggest doing so, if possible, for clarity

File: community/bolt/src/docs/dev/serialization.asciidoc [hunk 7]
```diff
@@ -420,6 +422,46 @@ Path (signature=0x50) {
...
+
+- `S` must always consist of an even number of integers, or be empty
+- The first, third, ... integer in `S` has a range encompassed by (..,-1] and [1,..). These represent the relationships in the path
+- The second, fourth, ... integer in `S` has a range encompassed by [0,..). These represent the nodes in the path
+- By definition, the first node in the path will _always_ have an index of 0, so we exclude this from `S` upon transmission. The idea is to construct `S'` by prepending 0 to `S` on the completion of a successful transmission
+- Let a path `P` be given by the following transmitted sequence [1, 1, -2, 2]. It follows that the corresponding full sequence, `S'`, is given by [0, 1, 1, -2, 2]
+- The first integer in `S` (1) is the index in `R` corresponding to the first relationship in `P`
```
Comment (petraselmer) [inline community/bolt/src/docs/dev/serialization.asciidoc:442]: Make consistent with "key" or "index" - same for below...

File: community/bolt/src/docs/dev/serialization.asciidoc [hunk 8]
```diff
@@ -420,6 +422,46 @@ Path (signature=0x50) {
...
+The elements trasmitted would be as follows:
+
+- `N`: `(A)`, `(B)`, `\(C)`
+- `R`: `[:X]`, `[:Y]`, `[:Z]`
+- `S`: [1, 1, 2, 2, -3, 1, -1, 0]
+
+By definition, the following is also implied:
+
+- `S'`: [0, 1, 1, 2, 2, -3, 1, -1, 0]
+
```
Comment (petraselmer) [inline community/bolt/src/docs/dev/serialization.asciidoc:464]: What about the corner cases... such as zero-length paths? This is just to preempt questions from people about this...

File: community/bolt/src/docs/dev/serialization.asciidoc (hunk 7 above)
Comment (zhenlineo) [inline community/bolt/src/docs/dev/serialization.asciidoc:442]: We were trying to use `list` and `index` to make the doc align with the code we have for Path

```
Path (signature=0x50) {
    List<Node> nodes
    List<UnboundRelationship> relationships
    List<Integer> sequence
}
```

File: community/bolt/src/docs/dev/serialization.asciidoc (hunk 8 above)
Comment (zhenlineo) [inline community/bolt/src/docs/dev/serialization.asciidoc:464]: The idea is to provide a good example to cover most of the cases for a reader to understand path. I've added a zero-length path example. What else example do you think is also important to add? I feel two example already cover all situations discussed in the original docs.

**Pull request merged**